### PR TITLE
ADR-101: Backport documentation to v0.38.x-experimental

### LINKED ISSUE #1475: ADR-101 Backport documentation to v0.38.x
commits: 46a4cea 
 
We need to backport the documentation for how the API is supposed to be used. THe main reason for this backport is that there are subtle differences to what is done in main. Mainly the existing docs have to be edited to contain:
- [x] Information on how the pruning is now happening in the background at regular time intervals, vs synchronously between heights
- [x] The fact that currently there exists a gRPC endpoint in v0.38.x that has to have a different endpoint from the one the  user uses to configure the gRPC endpoints used by the data companion.

### PR DESCRIPTION
**Pull request by hvanz**
Closes #1475 

Manual backport of #1307 

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: docs/data-companion/intro.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+order: 1
+parent:
+    title: Introduction
+    order: 1
+---
+
+# Introduction
+
+A proposal was made in
+[ADR-101](https://github.com/cometbft/cometbft/blob/thane/adr-084-data-companion-pull-api/docs/architecture/adr-101-data-companion-pull-api.md)
```
Comment (cason) [inline docs/data-companion/intro.md:11]: This link is broken.

File: docs/data-companion/intro.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+By also using the new gRPC services, it's possible now to retrieve data from the node, such as `block` and `block results`
+in a more efficient way.
+
+The [gRPC services](./grpc.md) document provides practical information and insights that will guide you through the
+process of using these services in order to create a Data Companion service.
+
+Note that this version of CometBFT (v0.38) has a gRPC service
+(`rpc/grpc/api.go`) that is considerd legacy code and will be removed in future
+releases. We encourage users to use the new gRPC service instead. In case you
```
Comment (cason) [inline docs/data-companion/intro.md:30]: >  We encourage users to use the new gRPC service instead.

For? I mean, they are not equivalent.

File: docs/data-companion/intro.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+The new pruning service allows granular control of what can be pruned such as blocks and state, ABCI results (if enabled), block
+indexer data and transaction indexer data.
+
+By also using the new gRPC services, it's possible now to retrieve data from the node, such as `block` and `block results`
+in a more efficient way.
+
+The [gRPC services](./grpc.md) document provides practical information and insights that will guide you through the
+process of using these services in order to create a Data Companion service.
+
+Note that this version of CometBFT (v0.38) has a gRPC service
```
Comment (cason) [inline docs/data-companion/intro.md:28]:
```suggestion
Note that this version of CometBFT (v0.38) already includes a gRPC service
```

File: docs/data-companion/intro.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+(`rpc/grpc/api.go`) that is considerd legacy code and will be removed in future
+releases. We encourage users to use the new gRPC service instead. In case you
+need to use the legacy endpoints, they should have a different URL than the gRPC
```
Comment (cason) [inline docs/data-companion/intro.md:31]:
```suggestion
need to use the legacy gRPC endpoints, make sure that they have a different URL than the gRPC
```

File: docs/data-companion/grpc.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+```
+
+Add the address for the non-privileged (regular) services, for example:
+
+```
+# TCP or UNIX socket address for the RPC server to listen on. If not specified,
+# the gRPC server will be disabled.
+laddr = "tcp://0.0.0.0:26090"
+```
+
```
Comment (cason) [inline docs/data-companion/grpc.md:47]: The note from line 111 should come here, actually.

File: docs/data-companion/pruning.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+# Pruning data via the pruning service
+
+CometBFT employs a sophisticated pruning logic to eliminate unnecessary data and reduce storage requirements.
+
+This document covers use cases where the pruning process on a CometBFT node can be influenced via the Data Companion
+pruning service API.
+
+The pruning process runs asynchronously in the background based on the retain
+heights set either by the application or via the Data Companion API. Before the
+Data Companion API was introduced, that is, before v0.38, pruning could be
```
Comment (cason) [inline docs/data-companion/pruning.md:17]: For all effects, this is v0.38, right? Moreover, this is not enabled in the actual (not experimental) v0.38.

File: docs/data-companion/pruning.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+---
+
+# Pruning data via the pruning service
+
+CometBFT employs a sophisticated pruning logic to eliminate unnecessary data and reduce storage requirements.
+
+This document covers use cases where the pruning process on a CometBFT node can be influenced via the Data Companion
+pruning service API.
+
+The pruning process runs asynchronously in the background based on the retain
```
Comment (cason) [inline docs/data-companion/pruning.md:15]: I would start with something like:

"Starting from this release, ..."

File: docs/data-companion/pruning.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+heights set either by the application or via the Data Companion API. Before the
```
Comment (cason) [inline docs/data-companion/pruning.md:16]:
```suggestion
heights set either by the application or via the Data Companion pruning service API. Before the
```

File: docs/data-companion/pruning.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+Data Companion API was introduced, that is, before v0.38, pruning could be
+triggered only by the application at some height, and CometBFT would not advance
+to the next height until the blocks were pruned. Now, the application just need
```
Comment (cason) [inline docs/data-companion/pruning.md:19]:
```suggestion
to the next height until the blocks were pruned. The pruning service was therefore synchronous. Now, the application just need
```

File: docs/data-companion/pruning.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+to set the desired retain heights, and then consensus and pruning run in
```
Comment (cason) [inline docs/data-companion/pruning.md:20]:
```suggestion
to set the desired retain heights, and the pruning service runs in parallel with consensus.
```

**Review by cason — COMMENTED** [review]
Some suggestions to improve the content added by this PR, not present in the original (backported) PR.

File: docs/data-companion/intro.md (hunk 2 above)
Comment (jmalicevic) [reply docs/data-companion/intro.md:30]: Yes, clarifying that the legacy gRPC service is used only for broadcasting transactions.

File: docs/data-companion/pruning.md (hunk 6 above)
Comment (jmalicevic) [reply docs/data-companion/pruning.md:17]:
```suggestion
Data Companion API was introduced, that is, in v0.38 and earlier versions, pruning could be
```

File: docs/data-companion/pruning.md (hunk 9 above)
Comment (jmalicevic) [reply docs/data-companion/pruning.md:19]:
```suggestion
to the next height until the blocks were pruned. In this experimental version of v0.38.x, the application just need
```

File: docs/data-companion/intro.md (hunk 1 above)
Comment (hvanz) [reply docs/data-companion/intro.md]: Fixed.

File: docs/data-companion/intro.md (hunk 2 above)
Comment (hvanz) [reply docs/data-companion/intro.md]: Right, I removed that line.

File: docs/data-companion/grpc.md (hunk 5 above)
Comment (hvanz) [reply docs/data-companion/grpc.md:47]: Done.

**Comment by hvanz** [comment]
Thanks @jmalicevic @cason for the suggestions!

File: docs/data-companion/pruning.md [hunk 11]
```diff
@@ -12,13 +12,15 @@ CometBFT employs a sophisticated pruning logic to eliminate unnecessary data and
...
-to the next height until the blocks were pruned. Now, the application just need
-to set the desired retain heights, and then consensus and pruning run in
-parallel. 
+Starting from this release, the pruning process runs asynchronously in the
+background based on the retain heights set either by the application or via the
+Data Companion pruning service API. Before the Data Companion API was
+introduced, that is, in v0.38 and earlier versions, pruning could be triggered
+only by the application at some height, and CometBFT would not advance to the
+next height until the blocks were pruned. The pruning service was therefore
+synchronous. In this experimental version of v0.38.x, the application just need
```
Comment (jmalicevic) [inline docs/data-companion/pruning.md:21]:
```suggestion
synchronous. In this experimental version of v0.38.x, the application just needs
```

**Review by jmalicevic — APPROVED** [review]
Minor typo, LGTM.

**Review by cason — APPROVED** [review]
Good.

**Pull request merged**